File: leet-code/543. Diameter of Binary Tree/DiameterOfBinaryTree543.java
public class DiameterOfBinaryTree543 {
    public static int maxDiameter = 1;

    public int diameterOfBinaryTree(TreeNode root) {
        int result = traverse(root);
        int answer = maxDiameter - 1;
        // maxDiameter 초기화
        maxDiameter = 1;
        return answer;
    }

    public int traverse(TreeNode root) {
        if (root == null) return 0;
        int left = traverse(root.left);
        int right = traverse(root.right);
        // 현재 정점까지의 최대 diameter 는 max + 1 + min
        maxDiameter = Math.max(maxDiameter, Math.max(left, right) + 1 + Math.min(left, right));
        // 현재 정점의 최대 깊이는 max + 1
        return Math.max(left, right) + 1;
    }
}

File: leet-code/543. Diameter of Binary Tree/DiameterOfBinaryTree543.kt
import kotlin.math.max

class DiameterOfBinaryTree543_2 {
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        var maxDiameter = 1

        fun dfs(node: TreeNode?): Int {
            if (node == null) return 0
            val left = dfs(node.left)
            val right = dfs(node.right)
            maxDiameter = max(maxDiameter, left + right + 1)
            return max(left, right) + 1
        }

        dfs(root)
        return maxDiameter - 1
    }
}
